### PR TITLE
Add disabled-features flag to ContourDeployment for provisioner

### PR DESCRIPTION
**Pull request by lubronzhan**
Fix https://github.com/projectcontour/contour/issues/5276

Based on previous art from @padlar https://github.com/projectcontour/contour/pull/5331.

This PR introduce support for disabled-features in provisioner so that any Contour instances associated with this ContourDeployment with disabled-features flag will stop reconciling CRDs defined inside.

Right now supported values are: `grpcroutes;tlsroutes;extensionservices;backendtlspolicies`

TODO:
- [x]  Add e2e test

### CODE REVIEW

File: internal/provisioner/model/model.go [hunk 1]
```diff
@@ -254,6 +255,18 @@ type ContourSpec struct {
 	// to only watch these set of namespaces
 	// default is nil, contour will watch resource of all namespaces
 	WatchNamespaces []string
```
Comment (lubronzhan) [inline internal/provisioner/model/model.go:257]: Will change this to original v1.Namespace type in a follow up PR.
I think it make more sense to keep model and types consistent, and put conversion function inside model

File: internal/provisioner/objects/deployment/deployment_test.go [hunk 2]
```diff
@@ -270,3 +271,34 @@ func TestNodePlacementDeployment(t *testing.T) {
...
 }
+
+func TestDesiredDeploymentWhenSettingDisabledFeature(t *testing.T) {
+	testCases := []struct {
+		description      string
+		disabledFeatures []contour_v1.Feature
+	}{
+		{
+			description:      "several disabled features",
+			disabledFeatures: []contour_v1.Feature{"ns1", "ns2"},
```
Comment (sunjayBhatia) [inline internal/provisioner/objects/deployment/deployment_test.go:282]: nit: use a feature name here just for readability (functionally doesnt matter and could by any string, but just for someone reading it later)

**Review by sunjayBhatia — COMMENTED** [review]
just a few comments for now, will do another pass shortly

File: test/e2e/provisioner/provisioner_test.go [hunk 3]
```diff
@@ -34,6 +34,7 @@ import (
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/utils/ptr"
```
[the remark this answers is not in the document]
Comment (sunjayBhatia) [reply test/e2e/provisioner/provisioner_test.go:37]: can do in a later issue of course cc @projectcontour/maintainers

[the remark this answers is not in the document]
Comment (lubronzhan) [reply test/e2e/provisioner/provisioner_test.go:37]: Cool, let me create a Jira for that and I can work on it as well

File: internal/provisioner/model/model.go (hunk 1 above)
Comment (lubronzhan) [reply internal/provisioner/model/model.go]: Cool let me change it here

**Comment by lubronzhan** [comment]
The upgrade test has many
>2024/02/08 08:34:22.321719 error making request #79: Get "http://172.18.255.200/": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
Is that related to something known issue

**Review by sunjayBhatia — APPROVED** [review]
I'll wait to merge in case anyone else wants a look but LGTM

File: apis/projectcontour/v1alpha1/contourdeployment.go [hunk 4]
```diff
@@ -131,6 +131,14 @@ type ContourSettings struct {
 	// +kubebuilder:validation:MinItems=1
 	// +kubebuilder:validation:MaxItems=42
 	WatchNamespaces []contour_api_v1.Namespace `json:"watchNamespaces,omitempty"`
+
+	// DisabledFeatures defines an array of Gateway API CRDs that will be ignored by
```
Comment (skriss) [inline apis/projectcontour/v1alpha1/contourdeployment.go:135]: Technically they are not all Gateway API resources (`extensionservices` is a Contour resource):
```suggestion
	// DisabledFeatures defines an array of resources that will be ignored by
```

**Review by skriss — CHANGES_REQUESTED** [review]
Just one docs nit, otherwise LGTM

@sunjayBhatia I'd be fine getting this in for the upcoming release, WDYT?

**Comment by sunjayBhatia** [comment]
> The upgrade test has many context deadline exceeded issue, is that related to something known issue
> 
> > 2024/02/08 08:34:22.321719 error making request #79: Get "http://172.18.255.200/": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
> 
> Didn't find the button to rerun the test

yeah, 

> Just one docs nit, otherwise LGTM
> 
> @sunjayBhatia I'd be fine getting this in for the upcoming release, WDYT?

yep sounds good to me 👍🏽

**Comment by skriss** [comment]
> > The upgrade test has many context deadline exceeded issue, is that related to something known issue
> > > 2024/02/08 08:34:22.321719 error making request #79: Get "http://172.18.255.200/": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
> > 
> > 
> > Didn't find the button to rerun the test
> 
> yeah, I dont think we made an issue for it but we've seen this a lot I believe since we haven't properly coordinated the Envoy/Contour Service/Deployment redeploy order in the provisioner

ref. https://github.com/projectcontour/contour/issues/5375, got closed as stale but still valid to do

**Review by skriss — APPROVED** [review]

**Pull request merged**